### DISCUSSION TITLE
btagging SFs in columnflow.production.cms

### DISCUSSION THREAD

**Discussion by JulesVandenbroeck** [columnflow/columnflow#417] [Ideas]
Hi everyone,

We (Ghent university team) were looking into the btagging scale factors and how they were applied in columnflow (see [btag.py](https://github.com/columnflow/columnflow/blob/master/columnflow/production/cms/btag.py)) and saw the weights are not following the fixed wp [guidelines from BTV](https://btv-wiki.docs.cern.ch/PerformanceCalibration/fixedWPSFRecommendations/#b-tagging-efficiencies-in-simulation). Is there a reason why the fixed wp b-tagging efficiency is not included in this calculation and is there a more recent version or an analysis that includes this or should we best write it ourselves (and share in case it is useful to columnflow?

**Comment by pkausw** [comment]
Hi @JulesVandenbroeck ,

apologies for the late reply - there was some offline discussion with CMS BTV expert Jan van der Linden, who is also in your group so I lost track of this point for a bit. The cms.btag weight producer targets the iterative shape btagging weights, which are different from the fixed working point SFs that are discribed in the reference you've linked above. I have verified again that the weights for the btag discriminant shape are implemented correctly.

Nevertheless, it would also be nice to implement some kind of support for fixed btag working points. However, I guess that this will need some additional work since there is a lot of input that the user needs to provide (e.g. the dataset/processes at gen level that are required to estimate the selection efficiency, which of course depends on the selections that are applied in an individual analysis). I'm moving this to the idea section since this will be a new project and is somewhat unrelated to the currently-implemented btag weights.

If you have started this work, feel free to let us know! Maybe we can work together to implement a general solution for multiple users :+1:

**Reply by JulesVandenbroeck** [reply]
Hi Philip,

Sorry for the very very late reply - I have put some effort into making a working implementation of a btagging scale factor for fixed working point ([see here my analysis gitlab](https://gitlab.cern.ch/ghentanalysis/columnflowanalysis/ttz/-/tree/master?ref_type=heads)). The implementation is two-fold:

1) calculate b-tagging efficiency with a custom task after `cf.MergeReducedEvents` ([see code](https://gitlab.cern.ch/ghentanalysis/columnflowanalysis/ttz/-/blob/master/ttz/tasks/studies/btagefficiency.py?ref_type=heads))
2) produce b-tagging scale factors with a producer which still needs work to correctly include all systematics  ([see code](https://gitlab.cern.ch/ghentanalysis/columnflowanalysis/ttz/-/blob/master/ttz/production/btag_weights.py?ref_type=heads)) 

The code currently requires the input of some additions to the config ([see code](https://gitlab.cern.ch/ghentanalysis/columnflowanalysis/ttz/-/blob/master/ttz/config/variables.py?ref_type=heads)):
- `config.x.default_btagAlgorithm` which provides "algorithm used, working point, and the discriminator in NanoAOD (example:` ("deepJet", "M", "btagDeepFlavB")`)
-  variables in which to bin the b-tagging efficiency ([see btv recommendations](https://btv-wiki.docs.cern.ch/PerformanceCalibration/fixedWPSFRecommendations/))
- `config.x.btag_dataset_groups` which give groups of datasets which can be combined for the b-tagging efficiency calculation

If there are any comments on the code that could be improved or cf tools I am unaware of that could be usefull please let me know. For general use,  I believe this implementation could be used for any analysis that uses a "basic" 1 working point for b-tagging but there are still some improvements to be made before it can be used by others I think.

I will make sure to give an update once it is finalized or if you would have comments.
Kind Regards,
Jules
